File: lab3_sort.hs
--Assignment 3- lab3.hs
--Wilfredo Mendivil
--017471907
--Due Date: 4/7/2020 at 9:30 

--Quicksort function. 
quickSort :: Ord a => [a] -> [a]  --use Ord to compare something type a. use a list to compare and return a list
quickSort [] = [] --initial parameter
quickSort (x:xs) = quickSort [y | y <- xs, y <= x] ++ [x] ++ quickSort [y | y <- xs, y > x] --quick sort both sides. if y is les than or equal to x it will be add x to the end and followed by the restof the list


data  Person = Person String Int deriving (Show,Eq) --person class holds a string and an int
name :: Person -> String --name is the string
name    (Person n a) = n --name is the first parameter

age :: Person ->Int --age is the int
age     (Person n a) = a --age is the second parameter

instance Ord Person where --goes through instances of person to be used in quick
    compare x y =  --compares 2 values
        let 
            i = compare (name x) (name y) --compare the name of two objects 
        in      
            if i == EQ then compare (age x) (age y) --if name is equal compare age
            else i  --otherwise compare name

data Person2 = Person2 String Int deriving (Show, Eq) --second person class
name2 :: Person2 -> String--name is the string
name2     (Person2 n a) = n--name is the first parameter

age2  :: Person2 ->Int --age is the int
age2     (Person2 n a) = a --age is the second parameter

instance Ord Person2 where --looks through instances where person2 is being compared
    compare x y = --compare 2 types
        let
		    i = flip compare (age2 x) (age2 y) --compare the age of two objects 
		in
		    if i == EQ then compare (name2 x) (name2 y)--if age is equal compare name
			else i   --otherwise compare age

          
            

main = do 
    putStrLn " list number after quickSort : " --print 
    print (quickSort [645.32, 37.40, 76.30, 5.40, -34.23, 1.11, -34.94, 23.37, 635.46, -876.22, 467.73, 62.26]) --print quicksorted values of floats in list
  
    putStrLn " Person sorted by Name" -- print message to screen
    print(quickSort[ Person "Hal" 20, --quicksort and print person list to screen sorted alphabetically
                              Person "Susann"31, 
                              Person "Dwight"19,
                              Person "Kassandra" 21,
                              Person "Lawrence" 25, 
                              Person "Cindy" 22,
                              Person "Cory" 27,
                              Person "Mac" 19,
                              Person "Romana" 27,
                              Person "Doretha" 32,
                              Person "Danna" 20,
                              Person "Zara" 23,
                              Person "Rosalyn" 26,
                              Person "Risa" 24,
                              Person "Benny" 28,
                              Person "Juan" 33,
                              Person "Natalie" 25])
							  
    putStrLn " Person sorted by Age"--print message to screen
    print(quickSort[ Person2 "Hal" 20,--quicksort and print person2 list to screen sorted numerically decending
                              Person2 "Susann"31,
                              Person2 "Dwight"19,
                              Person2 "Kassandra" 21,
                              Person2 "Lawrence" 25,
                              Person2 "Cindy" 22,
                              Person2 "Cory" 27,
                              Person2 "Mac" 19,
                              Person2 "Romana" 27,
                              Person2 "Doretha" 32,
                              Person2 "Danna" 20,
                              Person2 "Zara" 23,
                              Person2 "Rosalyn" 26,
                              Person2 "Risa" 24,
                              Person2 "Benny" 28,
                              Person2 "Juan" 33,
                              Person2 "Natalie" 25])

File: lab3_Sort.c
/*
Wilfredo Mendivil
CECS 424 Assignment 3- lab3.c
Due date : 4 / 7 / 2020
*/


#include <stdio.h>
#include <string.h>
#include <stdlib.h>

//struct person
struct Person
{
	char* name; //char pointer to name of person
	int age; //integer value holding the age of person
};

//compares two values type void*
int compare_float(void* a, void* b)
{
	return ((*(float*)a) - (*(float*)b));
}

//compare name of person
int comp_person_name(const void* a, const void* b)
{
	struct Person* temp_a = (struct Person*)a; //temp value holding Person a
	struct Person* temp_b = (struct Person*)b;//temp value holding Person b

	// Sort by name first then age
	if (strcmp(temp_a->name, temp_b->name) == 0) //compares two strings pointed to
		return (temp_a->age - temp_b->age); //if name is equal compare age

	return strcmp(temp_a->name, temp_b->name);//otherwise compare name
}
//compare age of person
int comp_person_age(const void* a, const void* b)
{
	struct Person* temp_a = (struct Person*)a;//temp value holding Person a
	struct Person* temp_b = (struct Person*)b;//temp value holding Person b

	// Sort by age first then name
	if ((temp_b->age - temp_a->age) == 0) //check if age is equal
		return (strcmp(temp_a->name, temp_b->name)); // compare by name if age is equal
	return (temp_b->age - temp_a->age);//otherwise compare by age
}
//print float array
void print_float_array(float* arr, int size)
{
	for (int i = 0; i < size; i++) {//itterate through aray
		printf("%.2f, ", arr[i]);//print formatted
	}
	printf("\n");//print empty line after loop ends
}
//print person array
void print_person_array(struct Person* arr, int size)
{
	for (int i = 0; i < size; i++) {//itterate through person array
		printf("%s, %d\n", arr[i].name, arr[i].age);//print formated person
	}
	printf("\n");//print empty line after loop ends
}

int main()
{
	// Sorting array of integers
	float numbers[] = { 645.32,37.40,76.30,5.40,-34.23,1.11,-34.04,23.37,635.46,-876.22,467.73,62.26 };
	int numbers_size = sizeof(numbers) / sizeof(numbers[0]);//size of numbers array by retrieving the byte size of numbers and dividing by a value inside
	qsort(numbers, numbers_size, sizeof(int), compare_float);//call quicksort with specific generic compare function
	printf("sorted float array is:\n");//print statement
	print_float_array(numbers, numbers_size);//print float array 
	printf("\n");//print empty line

	// Sorting array of structures
	struct Person people[] = {
		[0] .name = "Hal",[0].age = 20,
		[1].name = "Susann",[1].age = 31,
		[2].name = "Dwight",[2].age = 19,
		[3].name = "Kassandra",[3].age = 21,
		[4].name = "Lawrence",[4].age = 25,
		[5].name = "Cindy",[5].age = 22,
		[6].name = "Cory",[6].age = 27,
		[7].name = "Mac",[7].age = 19,
		[8].name = "Romana",[8].age = 27,
		[9].name = "Doretha",[9].age = 32,
		[10].name = "Danna",[10].age = 20,
		[11].name = "Zara",[11].age = 23,
		[12].name = "Rosalyn",[12].age = 26,
		[13].name = "Risa",[13].age = 24,
		[14].name = "Benny",[14].age = 28,
		[15].name = "Juan",[15].age = 33,
		[16].name = "Natalie",[16].age = 25
	};

	int person_size = sizeof(people) / sizeof(people[0]);//retrieve length of person array by getting the byte size of the array and dividing by the byte size of 1 value in the array
	qsort(people, person_size, sizeof(struct Person), comp_person_name);//call quick sort with specific compare person name
	printf("Sorted People by Name:\n");//print statement
	print_person_array(people, person_size);//calls method to print people arraylist

	qsort(people, person_size, sizeof(struct Person), comp_person_age);// call quick sort with specific compare person age
	printf("Sorted People by Age:\n");//print statement
	print_person_array(people, person_size);//calls method to print people arraylist

}